src/main/java/io/cryostat/configuration/CredentialsManager.java

```diff
@@ -233,19 +229,20 @@ public Set<ServiceRef> resolveMatchingTargets(int id) {
         return matchedTargets;
     }
 
-    public Set<ServiceRef> resolveMatchingTargets(String matchExpression) {
-        Set<ServiceRef> matchedTargets = new HashSet<>();
-        for (ServiceRef target : platformClient.listDiscoverableServices()) {
-            try {
-                if (matchExpressionEvaluator.get().applies(matchExpression, target)) {
-                    matchedTargets.add(target);
-                }
-            } catch (ScriptException e) {
-                logger.error(e);
-                break;
-            }
+    public Set<ServiceRef> resolveMatchingTargets(int id) {
+        Optional<String> matchExpression = get(id);
+        if (matchExpression.isEmpty()) {
+            return Set.of();
         }
-        return matchedTargets;
+        return getMatchedTargets(matchExpression.get(), platformClient.listDiscoverableServices());
+    }
+
+    public Set<ServiceRef> resolveMatchingTargets(String matchExpression) {
+        return getMatchedTargets(matchExpression, platformClient.listDiscoverableServices());
+    }
+
+    public Set<ServiceRef> resolveMatchingUniqueTargets(String matchExpression) {
```

Comment:
I'm not actually using this anywhere, I was going to use it for the `CredentialsListener` on `ADDED`, so that when credentials are added, we still check the db for the "most significant" target if there are multiple discoverable targets with the same jvmId. But this has a problem in that the `ADDED` event is also listened by the `DiscoveryStorage` to update the DiscoveryTree with the new non-null jvmId so it becomes a race condition. Would it make sense to try and maybe use the vertx event bus to notify the Credentials listener on the `RuleProcessor` when the db is finished persisting? Or maybe another way?

Reply:
> CredentialsListener on ADDED, so that when credentials are added, we still check the db for the "most significant" target if there are multiple discoverable targets with the same jvmId

The only difference this new method would make is in the case where there are multiple definitions for the same target (same ID), *and* the credential match expression matches both of them, I think? If so then I'm not sure how much this is worth worrying about.

> Would it make sense to try and maybe use the vertx event bus to notify the Credentials listener on the RuleProcessor when the db is finished persisting? Or maybe another way?

That would probably work but I would rather try to find some solution that doesn't involve adding more hooks to components and cascading events like this.

Reply:
Yeah, this is a small edge case where you would need identical targets saved, but no credentials for either, and after having a rule that matches both and saving a matching credential for both, the only thing that would happen is that it would start the same recording twice, which doesn't log an error it seems. Seems like JFR just overwrites the active recording so it's not a big issue.

Reply:
Yea, I think this is a small enough edge case to just not really worry about unless we find a scenario where it actually breaks something.